src/app/qgisapp.cpp

```diff
@@ -10001,7 +10001,23 @@ void QgisApp::reshapeFeatures()
 
 void QgisApp::addFeature()
 {
-  mMapCanvas->setMapTool( mMapTools->mapTool( QgsAppMapTools::AddFeature ) );
+  if ( mMapCanvas->mapTool()->toolName() != tr( "Add feature" ) )
+  {
+    mMapCanvas->setMapTool( mMapTools->mapTool( QgsAppMapTools::AddFeature ) );
+    if ( !mMapCanvas->currentLayer()->isSpatial() )
```

Comment:
This doesn't look very clean to me. Wouldn't we be better just NOT setting the map tool to add feature?